[Jangbagoony 이노 & MJ] db 스키마 작성, API 구현

eNoLJ opened the pull request:
안녕하세요. Jangbagoony팀에 이노, MJ입니다.

현재 스키마를 작성하고 기본적으로 제공되는 API 6개를 구현했습니다. 클라이언트에 최대한 빠르게 API를 제공하기 위해서 새로운 방식으로 API를 만들기보단 안정적인 방법으로 만들다보니 스키마에서는 1:다, 다:다와 같이 관계가 설정되어 있지만 코드 상으로는 매핑하여 사용하지 못했습니다. 이 부분은 추후 공부하여 하나씩 적용할 예정입니다.

지난 프로젝트와는 달리 스키마의 구조가 복잡해졌기 때문에 코드가 너저분한 느낌입니다. 하지만 평소와 마찬가지로 어떤식으로 개선을 해나가야할지 잘 떠오르지가 않네요.
이번에도 많은 조언 부탁드리겠습니다!!
잘부탁드려요ㅎㅎ

클라이언트 배포: http://15.164.68.136:80
서버 배포: http://15.164.68.136:8080
API 문서: https://app.gitbook.com/@todolist-2/s/side-dish

skawnkk committed:
Chore: 환경셋팅 및 폴더링

skawnkk committed:
Feat. MainMenu 마크업, itemCard 샘플 마크업#9

skawnkk committed:
Chore. 샘플카드 작성 #9

skawnkk committed:
Merge pull request #19 from skawnkk/fe/feature/mainmenu

샘플카드 및 theme 작업 중

eNoLJ committed:
Chore : .gitignore 파일 생성

eNoLJ committed:
Chore : 프로젝트 폴더인 BE 폴더 안에 .gitignore 파일 생성

eNoLJ committed:
Chore : side dish 스프링 부트를 이용한 gradle 프로젝트 생성

eNoLJ committed:
Feat : jdbc를 이용한 mysql 연결을 위해 application.properties 설정

eNoLJ committed:
Feat : side dish 프로젝트의 schema.sql 파일 생성

eNoLJ committed:
Chore: .gitignore 추가

skawnkk committed:
Merge pull request #20 from skawnkk/dev-fe

sample- hover 작업 중 #9

eNoLJ committed:
Feat: dish 관련 클래스와 레포지토리 생성

기본 프로젝트 골격을 잡기 위해 Dish, DishController, DishRepository, DishService, BestDishResponseDTO의 틀 생성

eNoLJ committed:
Feat: badge 관련 클래스와 레포지토리 생성

기본 프로젝트 골격을 잡기 위해 Badge, BadgeRepository 생성

Commit:
fix: image table 내 url 속성 변경

varchar(45) -> varchar(255)
url 삽입 불가에 따른 데이터 크기 키우기

Commit:
feat: 초기 데이터 삽입 쿼리 구현

eNoLJ committed:
Feat: DishBadge 관련 클래스와 레포지토리 생성

api 테스트를 위해 DishBadge, DishBadgeRepository 생성

eNoLJ committed:
Feat: Image 관련 클래스와 레포지토리 생성

api 테스트를 위해 Image, ImageRepository 생성

eNoLJ committed:
Feat: Badge 클래스에 getter 메소드 생성, BadgeRepository 인터페이스 설정

eNoLJ committed:
Feat: 필요한 데이터를 받아오기 위해 MainDishResponseDTO 클래스에 생성자 생성

eNoLJ committed:
Feat: Dish 클래스의 멤버변수를 꺼내기 위해 getter 메소드 생성

eNoLJ committed:
Feat: DishController 클래스에 모든 메인 반찬을 보여주기 위해 viewBestDishes 메소드 생성

eNoLJ committed:
Feat: DishRepository 인터페이스에 findAllByMenuCategoryId 메소드 생성

eNoLJ committed:
Feat: DishService 클래스에 메소드 생성

모든 메인 음식을 보여주기 위해 findAllMainDish, getImage, getBadges 메소드 생성

eNoLJ committed:
Merge pull request #25 from skawnkk/be/feature/setting

스프링 프로젝트 생성, jdbc 설정

eNoLJ committed:
Revert "스프링 프로젝트 생성, jdbc 설정"

eNoLJ committed:
Merge pull request #26 from skawnkk/revert-25-be/feature/setting

스프링 프로젝트 생성, jdbc 설정

eNoLJ committed:
Delete BE directory

eNoLJ committed:
Delete .gitignore

eNoLJ committed:
Revert "스프링 프로젝트 생성, jdbc 설정"

eNoLJ committed:
Merge pull request #27 from skawnkk/revert-26-revert-25-be/feature/setting

스프링 프로젝트 생성, jdbc 설정

eNoLJ committed:
chore: 충돌 해결

eNoLJ committed:
chore: 충돌 해결

eNoLJ committed:
chore: 충돌 해결

eNoLJ committed:
chore: 충돌 해결

eNoLJ committed:
Merge pull request #30 from skawnkk/be/feature/data

충돌을 바로잡는 pr

eNoLJ committed:
Merge branch 'be/feature/apiTest' of https://github.com/skawnkk/sidedish into be/feature/apiTest

eNoLJ committed:
Merge pull request #31 from skawnkk/be/feature/apiTest

main 음식에 대한 테스트 api 생성

eNoLJ committed:
Merge branch 'dev-be' of https://github.com/skawnkk/sidedish into dev-be

Commit:
feat: menu_category 별로 dish 생성

Commit:
Fix: detail_image url 속성의 타입 변경

varchar(64) -> varchar(255)

Commit:
Feat: image, detail_image, menu_category 데이터 생성 쿼리 추가

MJbae committed:
chore: merge init data(insert queries)

Be/refactor/data add

eNoLJ committed:
Feat: dish 테이블의 스키마 수정

dish 테이블의 s_price와 point의 기본 값을 0으로 설정

eNoLJ committed:
Feat: RequestLoggingFilterConfig 생성

request 요청에 대한 정보를 기록하기 위해 RequestLoggingFilterConfig 생성, 설정

eNoLJ committed:
Feat: BestMenuCategory에 대한 클래스와 레포지토리 생성

BestMenuCategory에 대한 정보를 조작하기 위해 BestMenuCategory 클래스와 BestMenuCategoryRepository 인터페이스 생성

eNoLJ committed:
Feat: DetailImage에 대한 클래스와 레포지토리 생성

DetailImage 대한 정보를 조작하기 위해 에 클래스와 DetailImageRepository 인터페이스 생성

eNoLJ committed:
Feat: MenuCategory에 대한 클래스와 레포지토리 생성

MenuCategory에 대한 정보를 조작하기 위해 MenuCategory 클래스와 MenuCategoryRepository 인터페이스 생성

eNoLJ committed:
Feat: Dish 클래스에 생성자와 메소드 생성

Dish 클래스에 hash와 stock을 받는 생성자 생성, 주문을 처리하기 위한 order 메소드, 주문을 처리할 수 있는지 체크하는 orderCheck 메소드 생성

eNoLJ committed:
Feat: DishRepository 인터페이스에 메소드 생성

DishRepository 인터페이스에 hash 값으로 찾는 findByHash 메소드, 베스트 메뉴를 카테고리로 찾아 랜덤으로 3개 가져오는 findByBestMenuCategoryIdAndRandomAndLimit 메소드, limit 제한을 받아 랜덤으로 찾는 findByRandomAndLimit 메소드 생성

eNoLJ committed:
Feat: ImageRepository 인터페이스에 메소드 생성

ImageRepository 인터페이스에 dishId를 통해 찾는 findAllByDishId 메소드 생성

eNoLJ committed:
Feat: ErrorMessage 이넘 생성

프로젝트의 에러 메세지를 상수로 관리하기 위해 ErrorMessage 이넘 생성

eNoLJ committed:
Feat: exception 생성

엔티티를 찾을 수 없을 때 사용하기 위해 EntityNotFoundException 클래스, 주문을 실패 했을 때 사용하기 위해 OrderFailedException 클래스 생성

eNoLJ committed:
Feat: BadgeService 클래스 생성

Badge 엔티티와 DishBadge 엔티티의 로직을 처리하기 위해 BadgeService 클래스 생성

eNoLJ committed:
Feat: CategoryService 클래스 생성

MenuCategory 엔티티와 BestMenuCategory 엔티티의 로직을 처리하기 위해 CategoryService 클래스 생성

eNoLJ committed:
Feat: ImageService 클래스 생성

Image 엔티티와 DetailImage 엔티티의 로직을 처리하기 위해 CategoryService 클래스 생성

eNoLJ committed:
Feat: DishService 클래스에 메소드 생성

Dish 관련 요청을 처리하기 위해 DishService 생성, 모든 베스트 메뉴를 찾는 findAllBestDish, 베스트 메뉴 카테고리 id에 해당하는 메뉴 중 랜덤으로 3개를 찾는 findAllBestDishByCategoryIdAndRandom, main, soup, side 메뉴를 찾는 findAllDishByCategory, 메뉴의 디테일 정보를 찾는 getDetailMenu, 모든 메뉴에 대해 limit 만큼 램덤으로 찾는 findDishByRandomLimit, 주문을 처리하는 orderDish 메소드 생성

eNoLJ committed:
Feat: DishController 클래스에 메소드 생성

Dish 관련 요청을 처리하기 위해 DishController 생성, 모든 best 메뉴 요청에 대한 viewAllBestDishes, 카테고리 별 best 메뉴 요청에 대한 viewBestDishesByCategory, 카테고리 별 메뉴 요청에 대한 viewDishes, 메뉴의 상세 정보 요청에 대한 viewDetailDish, 랜덤 메뉴 요청에 대한 viewRecommendDish, 주문에 대한 orderDish 메소드 생성

eNoLJ committed:
Feat: api 요청에 대해 response DTO 생성

eNoLJ committed:
Feat: .gitignore에 파일 추가

eNoLJ committed:
Merge pull request #35 from skawnkk/be/feature/api

프로젝트에 대한 API 생성

Commit:
fix: 해쉬값의 고유성 보장

Commit:
refator: thumb image 추가

MJbae committed:
Merge pull request #39 from skawnkk/be/refactor/correct_data

Be/refactor/correct data

eNoLJ committed:
Fix: 음수의 수량으로 주문이 되던 문제 해결

Commit:
Fix: CORS 발생에 따라 일시적으로 모든 요청에 대한 허용하는 방법으로 수정

MJbae committed:
Merge pull request #43 from skawnkk/be/fix/CORS

Fix: CORS 발생에 따라 일시적으로 모든 요청에 대한 허용하는 방법으로 수정

Commit:
refactor: 배지 데이터 변경

MJbae committed:
Merge pull request #46 from skawnkk/be/refactor/insert_data

refactor: 배지 데이터 변경

wheejuni commented on BE/src/main/java/com/mj_eno/sidedish/domain/Image/Image.java:
```diff
@@ -0,0 +1,28 @@
+
+public class Image {
+
+    @Id
+    Long id;
+    String url;
+    boolean top;
+    int dishId;
+
```
접근제어자가 하나도 없네요. 인생....

eNoLJ replied:
@wheejuni 와........ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 몰랐습니다. 웃음밖에 안나오네요........
이건 진짜 반성하도록 하겠습니다. 죄송합니다ㅠㅠ
다른 부분도 빠르게 수정하도록 하겠습니다.

MJbae replied:
반성합니다... ㅎㅎ...

wheejuni commented on BE/src/main/java/com/mj_eno/sidedish/domain/dish/Dish.java:
```diff
@@ -0,0 +1,102 @@
+public class Dish {
+
+    @Id
+    Long id;
+    String hash;
+    String alt;
+    String title;
+    String description;
+    int nPrice;
+    int sPrice;
+    int point;
+    int stock;
+    String deliveryType;
+    String deliveryInfo;
+    String deliveryFee;
+    int menuCategoryId;
+    Integer bestMenuCategoryId;
```
`Dish` 안에서 굉장히 많은 필드가 관리되고 있는데요, `@Embeddable` 등을 이용해서 객체 분리를 시도해볼 수 있을 것 같습니다. 
데이터 필드가 많을때는 데이터의 갈래(?) 에 따라 분류하고 분리하는 시도를 해보실 것을 권해 드립니다.

예를 들어 가격 정보라는 갈래 안에서 `nPrice`, `sPrice` 가 분리될 수 있을 것 같고,
`deliveryType`, `deliveryInfo`, `deliveryFee` 등을 따로 분리해서 관리할 수 있을 것 같습니다.

MJbae replied:
코멘트에 따라 Dish 내에 멤버변수로 있던 부분을 Price와 Delivery 클래스를 선언해서 분리했습니다. 분리하니까 코드의 가독성면에서 많이 개선되었네요~

wheejuni commented on BE/src/main/java/com/mj_eno/sidedish/domain/dish/DishRepository.java:
```diff
@@ -0,0 +1,20 @@
+    @Query("SELECT * FROM dish WHERE dish.best_menu_category_id = :bestMenuCategoryId ORDER BY rand() LIMIT 3;")
+    List<Dish> findByBestMenuCategoryIdAndRandomAndLimit(Long bestMenuCategoryId);
+
+    @Query("SELECT * FROM dish ORDER BY rand() LIMIT :limit;")
```
`ORDER BY rand()` 는 그리 좋은 방법이 아닙니다.
랜덤 인덱스를 서버에서 획득한 후 하나씩 불러오는 방법으로 처리해주세요.

MJbae replied:
코멘트에 따라 관련 내용을 검색하니까 성능 상의 이슈가 있는 것을 확인했습니다. DB에서는 Dish의 id만 조회하고, 자바 코드 상에서 랜덤 id를 선택하여 Dish를 조회하도록 수정했습니다.

wheejuni commented on BE/src/main/java/com/mj_eno/sidedish/exception/OrderFailedException.java:
```diff
@@ -0,0 +1,16 @@
+        super(errorMessage);
+    }
+
+    public OrderFailedException(ErrorMessage errorMessage) {
+        super(errorMessage.getErrorMessage());
+    }
```
`EntityNotFoundException` 에서도 비슷한 구현을 하셨던데 이렇게 가면 `ErrorMessage` 를 분리해낸 이점을 별로 살리지 못하는 구현이 됩니다.
오히려 이 예외 클래스에서 `ErrorMessage` 를 필드로 관리하고 `Errormessage` 에 더 많은 정보를 담게 될, 그러니까 기능이 확장될 대비를 하시는 것이 좋겠습니다.

MJbae replied:
커스텀 예외 클래스에서 ErrorMessage를 멤버변수로 관리하도록 수정했습니다. 그런데 이렇게 수정한 것이 브라이언의 의도와 맞는지 궁금합니다. 더불어 예외 클래스에서 멤버변수로 ErrorMessage를 관리하는 것이 유지보수면에서 어떤 이점이 있는지 잘 이해가 되지 않습니다... 조금 더 구체적으로 설명해주시면 감사하겠습니다 :)

wheejuni commented on BE/src/main/java/com/mj_eno/sidedish/service/DishService.java:
```diff
@@ -0,0 +1,95 @@
+public class DishService {
+
+    public final DishRepository dishRepository;
+    public final ImageService imageService;
+    public final BadgeService badgeService;
+    public final CategoryService categoryService;
```
`Service` 에서 다른 `Service` 에 의존하는 상황은 가급적 없었으면 합니다.
차라리 직접적으로 `ImageRepository` , `BadgeRepository` 등을 주입받는 방향이 더 좋아 보이기는 합니다.

MJbae replied:
다른 Service를 의존하여 구현한 이유는 DishService를 보기 좋게 만들기 위해서 입니다. 물론 다른 Service를 의존함에 따라 순환참조가 발생할 가능성이 있지만 현재의 설계에서 그러한 이슈가 발생할 수 없다고 판단했습니다. 
처음에는 순환참조 이슈를 고려하여 7개의 Repository를 주입 받아서 구현했습니다. 그런데 DishService가 너무 비대해져 가독성이 떨어지는 코드가 되어서 위와 같이 수정하였습니다.

wheejuni commented on BE/src/main/java/com/mj_eno/sidedish/web/DishController.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.List;
+
+@RestController
+@CrossOrigin("*")
```
모든 요청에 대해 CORS를 허용해야 한다면 컨트롤러 이전에 `Filter` 에서 처리하는 방법을 고민해주세요.
_spring cors filter_, _spring mvc cors config_ 등의 검색어로 구글링해보시고 솔루션을 찾아 주세요.

MJbae replied:
filter에서 처리하면 controller마다 해당 어노테이션을 붙일 필요가 없겠군요!

모든 요청에 대해 CORS를 허용하는 부분은 프런트엔드 개발자의 편의를 위해 임시적으로 적용했습니다. 로컬에서 개발을 마치면 서버의 요청에 대해서만 허용하는 식으로 변경할 예정입니다.

MJbae replied:
localhost:3000과 서버의 요청에 대해서만 CORS를 허용하는 식으로 변경했습니다. 코멘트에 따라 filter에서 관련 설정을 하는 것으로 수정했습니다. 

참고로 localhost:3000은 프런트엔드 개발자의 편의를 위해 잠시 열어두었습니다.

wheejuni commented on BE/src/main/java/com/mj_eno/sidedish/web/DishController.java:
```diff
@@ -0,0 +1,57 @@
+
+    @GetMapping("/best/{categoryId}")
+    public BestDishResponseDTO viewBestDishesByCategory(@PathVariable Long categoryId) {
+        logger.info(categoryId + "번 카테고리의 best 메뉴 요청");
```
`logger` 에서는 문자열 템플릿을 사용하는 것이 좋다고 리뷰를 충분히 드렸었던 것 같습니다.

```suggestion
        logger.info("{}번 카테고리의 best 메뉴 요청", categoryId);

MJbae replied:
이 부분도 반성합니다... 더 신경 써야겠네요!

MJbae replied:
코멘트에 따라 템플릿에 맞춰서 수정했습니다~

wheejuni reviewed (changes requested):
코드가 너저분한 느낌이 든다면 일단 그것만으로도 더 좋은 코드를 작성할 준비가 된 것입니다. 💯 
그러나 거창한 리팩토링이나 구조 개선을 생각하기 이전에 우선 **기본적으로 지켜야 할 것들을 잘 지키는 일** 에서부터 리팩토링이 출발돼야 할 것입니다. 
반복적인 피드백을 통해 1월달부터 이 점을 항상 강조해 드리고 있습니다. 제 강조의 뜻이 잘 전달되지 않는 것 같아서 개인적으로는 반성을 많이 하고 있습니다. 
코딩의 재미를 느끼고 프로젝트의 공정률을 높이는 것도 의미있지만, 우리가 무엇을 위해 팀 프로젝트를 수행하고 코드 리뷰를 요청하고 있는지도 생각해볼 만 합니다. 
감사합니다.

eNoLJ committed:
Feat: 클래스 멤버변수에 접근제어자 설정

eNoLJ committed:
Refactor: exception 클래스에 ErrorMessage를 멤버변수로 갖도록 수정

Errormessage 에 더 많은 정보를 담게 되거나 기능이 확장될 때를 대비하여 exception 클래스에 ErrorMessage를 멤버변수로 갖도록 수정

eNoLJ committed:
Refactor: logger 메세지를 문자열 템플릿을 사용하도록 수정

eNoLJ committed:
Refactor: dish의 random 조회 로직 변경

변경 전: db에서 dish를 랜덤으로 조회
변경 후: db에서 dish의 id만 조회 + 코드상으로 램덤하게 id 선택 + id에 해당하는 dish 조회

eNoLJ committed:
Refactor: @Embedded를 사용하여 Dish 클래스 수정

수정 전: dish 클래스에 nPrice, sPrice, point, deliveryInfo, deliveryFee, deliveryType 멤버변수 존재
수정 후: Price, Delivery 클래스를 만들고 dish에 @Embedded를 사용하여 멤버변수 분리

Commit:
refactor: cors 적용을 config filter로 변경

MJbae committed:
Merge pull request #52 from skawnkk/be/refactor/cors

refactor: cors 적용을 config filter로 변경
